keps/sig-network/3836-kube-proxy-improved-ingress-connectivity-reliability/README.md

```diff
@@ -0,0 +1,423 @@
+   `healthz` state, this can then be used to correlate to networking metrics
+   surrounding new/established connections on the node. Ex: a failing `healthz`
+   should correlate to a total drop in the count of new connections and with a
+   zero-or-negative rate of established connections. E2E tests should also be
+   designed with this specific goal in mind, i.e: validating the impact of a
+   failing kube-proxy on ingress connectivity. Kube-proxy currently has a lot of
+   metrics regarding how its health is doing, but no direct red/green indicator
+   of what the end result of its health is. One such metric could be
+   `proxy_health_check_200_count` / `proxy_health_check_503_count`
+
+3. The feature could be disabled for user who is dependent upon such behavior by
+   means of flipping the feature flag to off.
+
+## Design Details
+
+1. Implement change in Kube-proxy which allows the healthcheck server access to
+   the `healthz` state.
+
+2. Implement Kube-proxy change invoking client-go's `SetWatchErrorHandler`  on
+   watch errors from the API server.
+
+3. Implement change in Kube-proxy which will react to changes on the Node object
+   and once the taint `ToBeDeletedByClusterAutoscaler` is placed on the Node
+   object: start failing it's `healthz` state.
+
+4. Write document to be published at: https://kubernetes.io/docs/concepts/ which
+   details: a) how determining node/instance health can best be done for
+   Kubernetes clusters b) how Kube-proxy will do it once the changes proposed in
+   this KEP are merged c) what some pitfalls with other methods might be.
+
+### Test Plan
+
+#### Prerequisite testing updates
+
+#### Unit tests
+
+Update the Kube-proxy unit tests to include the `healthz` answer for the
+healthcheck server test suite.
+
+#### Integration tests
+
+Kubernetes is mostly tested via unit tests and e2e, not integration, and this is
+not expected to change.
+
+#### e2e tests
+
+- Add E2E tests for connectivity to services on terminating nodes and
+  validate graceful termination of TCP connections.
+
+### Graduation Criteria
+
+#### Alpha
+
+- E2E tests coded before any feature implementation is made which highlights the
+  existing problem.
+- Feature implemented behind a feature flag.
+
+#### Beta
+
+- Document written at https://kubernetes.io/docs/concepts/
+- No issues reported.
+
+#### GA
+
+- No issues reported during two releases.
+
+### Upgrade / Downgrade Strategy
+
+Any upgrade to a version enabling the feature, succeeded by a downgrade to a
+version disabling it, is not expected to be impact ingress in any way, given
+that Kube-proxy is healthy on all cluster nodes. Should Kube-proxy not be
+healthy: then ingress for `eTP:Local` services, might be impacted on an upgrade
+since Kube-proxy would start failing the HC, whereas it wasn't before the
+upgrade. 
+
+### Version Skew Strategy
+
+This doesn't touch load balancer / HC API, so even though an old Kube-proxy
```

Comment:
If we change the URL, then clouds can't start using it until N+2 (soon N+3) releases, to ensure lagging nodes are updated to have the new URL

Reply:
I will not add changing the URL to this KEP, because of what I mentioned: https://github.com/kubernetes/enhancements/pull/3837#discussion_r1100642321